src/client/gui/gui_cmd.cpp

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "gui_cmd.h"
+#include "argparser.h"
+
+#include <multipass/cli/client_common.h>
+#include <multipass/cli/client_platform.h>
+#include <multipass/cli/format_utils.h>
+#include <multipass/version.h>
+
+#include <QCoreApplication>
+#include <QDesktopServices>
+#include <QtConcurrent/QtConcurrent>
+
+namespace mp = multipass;
+namespace cmd = multipass::cmd;
+using RpcMethod = mp::Rpc::Stub;
+using namespace std::chrono_literals;
+
+mp::ReturnCode cmd::GuiCmd::run(mp::ArgParser* parser)
+{
+    if (!QSystemTrayIcon::isSystemTrayAvailable())
+    {
+        cerr << "System tray not supported\n";
+        return ReturnCode::CommandFail;
+    }
+
+    create_actions();
+    create_menu();
+    tray_icon.show();
+
+    return static_cast<ReturnCode>(QCoreApplication::exec());
+}
+
+void cmd::GuiCmd::create_actions()
+{
+    retrieving_action = tray_icon_menu.addAction("Retrieving instances...");
+    about_separator = tray_icon_menu.addSeparator();
+    quit_action = tray_icon_menu.addAction("Quit");
+}
+
+void cmd::GuiCmd::update_menu()
+{
+    std::vector<std::string> instances_to_remove;
+    tray_icon_menu.removeAction(retrieving_action);
+
+    auto reply = list_future.result();
+
+    for (auto it = instances_entries.cbegin(); it != instances_entries.cend(); ++it)
+    {
+        auto instance = std::find_if(reply.instances().cbegin(), reply.instances().cend(),
+                                     [it](const ListVMInstance& instance) { return it->first == instance.name(); });
+
+        if (instance == reply.instances().cend())
+        {
+            instances_to_remove.push_back(it->first);
+        }
+    }
+
+    for (const auto instance : instances_to_remove)
+    {
+        instances_entries.erase(instance);
+    }
+
+    for (const auto& instance : reply.instances())
+    {
+        auto name = instance.name();
+        auto state = QString::fromStdString(mp::format::status_string_for(instance.instance_status()));
+
+        auto it = instances_entries.find(name);
+
+        if (it != instances_entries.end())
+        {
+            auto instance_state = it->second.state;
+            if (state == "DELETED")
```

Comment:
This diff changes status_string_for to replace "DELETED" with "Deleted" - so this branch will never be followed afaics

Reply:
It would be safer to do comparisons with the enum returned by instance.instance_status()

Reply:
Yeah, why didn't I just use the enum?????  /me shrugs